File: scripts/ci/support/AtlasVaultMaliciousServerClient.swift
import CryptoKit
import Foundation

private func load(_ path: String) throws -> [String: Any] {
  try JSONSerialization.jsonObject(with: Data(contentsOf: URL(fileURLWithPath: path)))
    as! [String: Any]
}
private func canonical(_ value: Any) throws -> Data {
  try JSONSerialization.data(
    withJSONObject: value, options: [.sortedKeys, .withoutEscapingSlashes])
}
private func digest(_ value: Any) throws -> String {
  SHA256.hash(data: try canonical(value)).map { String(format: "%02x", $0) }.joined()
}

@main
private struct AtlasVaultMaliciousServerClient {
  static func main() throws {
    let args = Array(CommandLine.arguments.dropFirst())
    let mode = args[0]
    let plan = try load(args[2])
    let context = plan["context"] as! [String: Any]
    var results = [String: Any]()
    for (name, scenario) in plan["scenarios"] as! [String: [String: Any]] {
      let root = URL(fileURLWithPath: args[1]).appendingPathComponent(name)
      try FileManager.default.createDirectory(at: root, withIntermediateDirectories: true)
      let key = Data(SHA256.hash(data: Data("c24-synthetic-client-storage".utf8)))
      let c = try AtlasVaultGuardedSyncState(
        fileURL: root.appendingPathComponent("accepted-recovery.state"), encryptionKey: key,
        accountID: context["account_id"] as! String, vaultID: context["vault_id"] as! String,
        collectionID: context["collection_id"] as! String,
        keyEpoch: (context["key_epoch"] as! NSNumber).int64Value,
        trustedSigner: Data(base64Encoded: plan["public_b64"] as! String)!)
      let inbox = try AtlasVaultDurableEncryptedInbox(
        fileURL: root.appendingPathComponent("inbox.state"), encryptionKey: key)
      let outbox = try AtlasVaultDurableEncryptedOutbox(
        fileURL: root.appendingPathComponent("outbox.state"), encryptionKey: key)
      if mode == "prepare" { try c.initialize() }
      var categories = [String]()
      let before = try c.checkpoint()
      let actions =
        mode == "prepare"
        ? scenario["baseline"] as! [[String: Any]]
        : mode == "attack" ? scenario["attack"] as! [[String: Any]] : []
      for action in actions {
        func interrupt(_ point: String) throws {
          if action["stop_after"] as? String == point {
            try canonical(["interrupted_after": point]).write(
              to: URL(fileURLWithPath: args[3]), options: .atomic)
            while true { Thread.sleep(forTimeInterval: 60) }
          }
        }
        do {
          if let path = action["peer"] as? String {
            let peer = try load(path)[name] as! [String: Any]
            _ = try c.compareEvidence(peer["history"] as! [[String: Any]])
            categories.append("COMPATIBLE_PREFIX_NOT_FRESHNESS")
          } else {
            let p = (plan["packets"] as! [String: [String: Any]])[action["packet"] as! String]!
            let view = p["view"] as! [String: Any]
            let accepted = try c.ingest(
              view: view, registry: p["registry"] as! [[String: Any]],
              collection: p["collection"] as! [String: Any],
              opaqueState: Data(base64Encoded: p["opaque_b64"] as! String)!)
            try interrupt("admission")
            let op = try AtlasVaultEncryptedPatchOperation(
              jsonObject: p["operation"] as! [String: Any])
            try outbox.enqueue(op)
            try interrupt("outbox")
            let pending = try inbox.pendingOperations()
            if !pending.isEmpty
              && (pending.count != 1
                || NSDictionary(dictionary: pending[0].jsonObject)
                  != NSDictionary(dictionary: op.jsonObject))
            {
              throw NSError(domain: "C24UnexpectedReceipt", code: 1)
            }
            if try inbox.cursor() != view["root"] as? String && pending.isEmpty {
              try inbox.stagePage(
                expectedCursor: inbox.cursor(), nextCursor: view["root"] as? String,
                operations: [op])
            }
            try interrupt("inbox")
            while try inbox.applyNext({ _ in }) != nil {}
            try interrupt("receipt")
            try outbox.confirmRemoteAcceptance(op.operationID)
            categories.append(accepted ? "ACCEPTED" : "IDEMPOTENT")
          }
        } catch let error as AtlasVaultSyncRecoveryError {
          categories.append(error.rawValue)
        }
      }
      var called = false
      do { try c.automaticSync { called = true } } catch let error as AtlasVaultSyncRecoveryError {
        guard error.rawValue == "ATLAS_RECOVERY_PENDING" else { throw error }
      }
      let checkpoint = try c.checkpoint()
      let recovery = try c.recovery()
      results[name] = [
        "before": before, "checkpoint": checkpoint, "recovery": recovery,
        "history": try c.exportEvidence(), "evidence": try c.evidence(),
        "categories": categories, "automatic_sync_fenced": !called,
        "cursor": try inbox.cursor() ?? NSNull(),
        "pending_outbox": try outbox.pendingOperations().count,
        "state_sha256": try digest(checkpoint), "recovery_sha256": try digest(recovery),
      ]
    }
    try canonical(results).write(to: URL(fileURLWithPath: args[3]), options: .atomic)
    if mode != "inspect" { while true { Thread.sleep(forTimeInterval: 60) } }
  }
}

File: apps/atlas_flutter/test/support/atlas_vault_malicious_server_client.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';
import 'package:atlas/src/atlas_vault/sync_queue.dart';
import 'package:cryptography/cryptography.dart';

Map<String, Object?> object(dynamic value) =>
    Map<String, Object?>.from(value as Map);
dynamic sorted(dynamic value) {
  if (value is Map) {
    final keys = value.keys.cast<String>().toList()..sort();
    return {for (final key in keys) key: sorted(value[key])};
  }
  if (value is List) return value.map(sorted).toList();
  return value;
}

List<int> canonical(dynamic value) => utf8.encode(jsonEncode(sorted(value)));
Future<String> digest(dynamic value) async => (await Sha256().hash(
  canonical(value),
)).bytes.map((b) => b.toRadixString(16).padLeft(2, '0')).join();

Future<void> main(List<String> args) async {
  final mode = args[0], directory = args[1];
  final plan = jsonDecode(File(args[2]).readAsStringSync()) as Map;
  final context = plan['context'] as Map, results = <String, Object?>{};
  for (final entry in (plan['scenarios'] as Map).entries) {
    final name = entry.key as String, scenario = entry.value as Map;
    final root = Directory('$directory/$name')..createSync(recursive: true);
    final key = Uint8List.fromList(
      (await Sha256().hash(utf8.encode('c24-synthetic-client-storage'))).bytes,
    );
    final c = AtlasVaultGuardedSyncState(
      file: File('${root.path}/accepted-recovery.state'),
      encryptionKey: key,
      accountId: context['account_id'] as String,
      vaultId: context['vault_id'] as String,
      collectionId: context['collection_id'] as String,
      keyEpoch: context['key_epoch'] as int,
      trustedSigner: base64Decode(plan['public_b64'] as String),
    );
    final inbox = AtlasVaultDurableEncryptedInbox(
      File('${root.path}/inbox.state'),
      encryptionKey: key,
    );
    final outbox = AtlasVaultDurableEncryptedOutbox(
      File('${root.path}/outbox.state'),
      encryptionKey: key,
    );
    if (mode == 'prepare') await c.initialize();
    final categories = <String>[], before = await c.checkpoint();
    final actions = mode == 'prepare'
        ? scenario['baseline'] as List
        : mode == 'attack'
        ? scenario['attack'] as List
        : [];
    for (final action in actions) {
      Future<void> interrupt(String point) async {
        if (action['stop_after'] == point) {
          File(args[3]).writeAsBytesSync(canonical({'interrupted_after': point}), flush: true);
          while (true) { await Future<void>.delayed(const Duration(seconds: 60)); }
        }
      }
      try {
        if (action['peer'] != null) {
          final peer =
              jsonDecode(File(action['peer'] as String).readAsStringSync())
                  as Map;
          await c.compareEvidence(
            (peer[name]['history'] as List).map(object).toList(),
          );
          categories.add('COMPATIBLE_PREFIX_NOT_FRESHNESS');
        } else {
          final p = plan['packets'][action['packet']] as Map;
          final accepted = await c.ingest(
            object(p['view']),
            (p['registry'] as List).map(object).toList(),
            object(p['collection']),
            base64Decode(p['opaque_b64'] as String),
          );
          await interrupt('admission');
          if (accepted) {
            final op = AtlasVaultEncryptedPatchOperation.fromJson(
              object(p['operation']),
            );
            await outbox.enqueue(op);
            await interrupt('outbox');
            await inbox.stagePage(
              expectedCursor: await inbox.readCursor(),
              nextCursor: p['view']['root'] as String,
              operations: [op],
            );
            await interrupt('inbox');
            while (await inbox.applyNext((_) async {}) != null) {}
            await interrupt('receipt');
            await outbox.confirmRemoteAcceptance(op.operationId);
          }
          categories.add(accepted ? 'ACCEPTED' : 'IDEMPOTENT');
        }
      } on AtlasVaultStateViewException catch (error) {
        categories.add(error.code);
      }
    }
    var called = false;
    try {
      await c.automaticSync(() async {
        called = true;
      });
    } on AtlasVaultStateViewException catch (error) {
      if (error.code != 'ATLAS_RECOVERY_PENDING') rethrow;
    }
    final checkpoint = await c.checkpoint(), recovery = await c.recovery();
    results[name] = {
      'before': before,
      'checkpoint': checkpoint,
      'recovery': recovery,
      'history': await c.exportEvidence(),
      'evidence': await c.evidence(),
      'categories': categories,
      'automatic_sync_fenced': !called,
      'cursor': await inbox.readCursor(),
      'pending_outbox': (await outbox.pendingOperations()).length,
      'state_sha256': await digest(checkpoint),
      'recovery_sha256': await digest(recovery),
    };
  }
  File(args[3]).writeAsBytesSync(canonical(results), flush: true);
  if (mode != 'inspect') {
    while (true) {
      await Future<void>.delayed(const Duration(seconds: 60));
    }
  }
}
